File: julia/day24_part1_2015.jl
using Printf

function main()
    data = read("input.txt", String)
    lines = split(strip(data), "\n")
    packages = parse.(Int, lines)
    totalWeight = sum(packages)
    targetWeight = div(totalWeight, 3)
    bestQE = typemax(Int)
    bestLength = typemax(Int)

    for comb in 1:(2^length(packages)-1)
        groupWeight, qe, groupLength = 0, 1, 0
        for i in 1:length(packages)
            if comb & (1 << (i-1)) != 0
                groupWeight += packages[i]
                qe *= packages[i]
                groupLength += 1
            end
        end
        if groupWeight == targetWeight && groupLength <= bestLength
            if groupLength < bestLength || qe < bestQE
                bestLength = groupLength
                bestQE = qe
            end
        end
    end

    @printf("%d\n", bestQE)
end

main()

File: javascript/day24_part1_2015.js
const fs = require('fs');

const data = fs.readFileSync('input.txt', 'utf8');
const lines = data.trim().split('\n');

const packages = lines.map(Number);
const totalWeight = packages.reduce((sum, weight) => sum + weight, 0);

const targetWeight = Math.floor(totalWeight / 3);
let bestQE = Number.MAX_SAFE_INTEGER;
let bestLength = Number.MAX_SAFE_INTEGER;

for (let comb = 1; comb < (1 << packages.length); comb++) {
  let groupWeight = 0, qe = 1, groupLength = 0;
  for (let i = 0; i < packages.length; i++) {
    if (comb & (1 << i)) {
      groupWeight += packages[i];
      qe *= packages[i];
      groupLength++;
    }
  }
  if (groupWeight === targetWeight && groupLength <= bestLength) {
    if (groupLength < bestLength || qe < bestQE) {
      bestLength = groupLength;
      bestQE = qe;
    }
  }
}

console.log(bestQE);